From e21bf491132493a4366c62baf3ab00ab55cc4694 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 17:28:02 -0400
Subject: [net] svcrdma: Fence LOCAL_INV work requests

Message-id: <2d917142f67a620b2ba742ffc3e7c07c8893d62d.1411578494.git.dledford@redhat.com>
Patchwork-id: 96667
O-Subject: [Patch RHEL7 09/26] svcrdma: Fence LOCAL_INV work requests
Bugzilla: 1107621
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Steve Wise <swise@opengridcomputing.com>

Fencing forces the invalidate to only happen after all prior send
work requests have been completed.

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Reported by : Devesh Sharma <Devesh.Sharma@Emulex.Com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>

(cherry picked from commit 83710fc753d2ae158aa3cb7a7966d9c1bd05b792)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/svc_rdma_recvfrom.c b/net/sunrpc/xprtrdma/svc_rdma_recvfrom.c
index 52d9f2c..8f92a61 100644
--- a/net/sunrpc/xprtrdma/svc_rdma_recvfrom.c
+++ b/net/sunrpc/xprtrdma/svc_rdma_recvfrom.c
@@ -338,7 +338,7 @@ static int rdma_read_chunk_frmr(struct svcxprt_rdma *xprt,
   memset(&inv_wr, 0, sizeof(inv_wr));
   inv_wr.wr_id = (unsigned long)ctxt;
   inv_wr.opcode = IB_WR_LOCAL_INV;
-  inv_wr.send_flags = IB_SEND_SIGNALED;
+  inv_wr.send_flags = IB_SEND_SIGNALED | IB_SEND_FENCE;
   inv_wr.ex.invalidate_rkey = frmr->mr->lkey;
  }
  ctxt->wr_op = read_wr.opcode;
-- 
1.7.1